### ISSUE TITLE
please retry downloads

### ISSUE THREAD

**Issue by luan** [concourse/s3-resource#15]
We keep getting errors like:

```
error running command: read tcp 10.254.0.22:54409->54.231.64.25:443: read: connection reset by peer
```

And have to manually press the re-run button because this is something we have no control over.

**Comment by robdimsdale** [comment]
+1 we are seeing this too. Can we do something like exponential backoff?

**Comment by vito** [comment]
As far as I can tell the AWS SDK already does this out of the box, and we shouldn't have to configure anything.

https://github.com/concourse/s3-resource/blob/master/s3client.go#L47-L58

https://github.com/aws/aws-sdk-go/blob/c09f1bb2f2270137bd1708d4668dc6beb2fb52f2/aws/service/service.go#L48

https://github.com/aws/aws-sdk-go/blob/7d9ac5f2079172b9725513796718eaad96218f3b/aws/service/default_retryer.go

https://github.com/aws/aws-sdk-go/blob/66c840e9981dd121a4239fc25e33b6c1c1caa781/aws/corehandlers/handlers.go#L90

**Comment by robdimsdale** [comment]
interesting. It looks like the default number of retries is 3 (https://github.com/aws/aws-sdk-go/blob/c09f1bb2f2270137bd1708d4668dc6beb2fb52f2/aws/service/service.go#L49) - can we add a way to configure this in the bosh-manifest?

**Comment by vito** [comment]
Would rather land on a sane default and prove that the current value (3) is what's wrong. Their implementation for retry logic is super generic and pretty hard to follow, maybe we're not triggering it after all.

**Comment by robdimsdale** [comment]
Ok. It doesn't seem unreasonable for this to be user-configurable though, as concourse might be deployed in infrastructure where the 'sane default' might not be enough.

Are you not seeing this issue anywhere? We're seeing this across pipelines (and across teams)

**Comment by vito** [comment]
We don't touch s3 until late in our pipeline (basically after shipping), so nope, not hitting it ourselves. I agree it should be configurable but I wouldn't consider that to be resolving the issue, since the default should be sane.

**Comment by rosenhouse** [comment]
I'm not convinced that this isn't a bug in aws-sdk-go.

Looking at [`IsErrorRetryable()`](https://github.com/aws/aws-sdk-go/blob/master/aws/request/retryer.go#L58) it seems that it will only retry errors that are `awserr.Error` types.  I don't see anywhere that a low-level TCP error would be converted to that type.  Furthermore, the [test coverage of the request retries](https://github.com/aws/aws-sdk-go/blob/master/aws/request/request_test.go) does not include any kind of TCP errors, only HTTP errors.

cc @robdimsdale

**Comment by rosenhouse** [comment]
Looking into this more, it does aws-sdk-go does retry TCP errors.  But [the back-off is fast](https://github.com/aws/aws-sdk-go/blob/fd136b7571693211815b5c780eebcb52c37a1b46/aws/client/default_retryer.go#L34).  Even by the 3rd retry, the delay is never more than 1/2 a second.

The other thing that may be interacting here is this issue: https://github.com/golang/go/issues/8946

On the SSL Service, we did 7 retries with 100ms base, giving a final delay of nearly 13 seconds.  For Concourse, it might make sense to something similar.
